### PR TITLE
General source map support

### LINKED ISSUE #37: Support transpiled JavaScript
We already support source-maps for coffeescript, but we haven't tested with transpiled JavaScript.

### PR DESCRIPTION
**Pull request by matthewloring**
Source maps are now discovered while traversing all javascript
files in the project allowing the discovery of js -> js
transpiled files by the same mechanism as other compiled
languages like coffeescript.

Fixes #37.

### CODE REVIEW

**Comment by matthewloring** [comment]
@ofrobots PTAL.

**Comment by matthewloring** [comment]
@ofrobots I've updated the source map reading to be fully async.

**Comment by matthewloring** [comment]
@ofrobots Added testing with source-maps that have multiple source files.

File: lib/scanner.js [hunk 1]
```diff
@@ -134,31 +144,56 @@ function findJSFiles(baseDir, callback) {
...
+      shasum.update(data);
+      hash = shasum.digest('hex');
+    }
+    var contents = data.toString();
+    var lines = (contents.match(/\n/g) || []).length;
+    var sourceMapLine;
+    var mapMatches = SOURCE_MAP_REGEX.exec(contents);
+    if (mapMatches && mapMatches.length > 1) {
+      sourceMapLine = mapMatches[1];
+    }
```
Comment (ofrobots) [inline lib/scanner.js]: Merge this `if` with the following `if` block.

File: lib/scanner.js [hunk 2]
```diff
@@ -134,31 +144,56 @@ function findJSFiles(baseDir, callback) {
...
+      var mapUrl = path.join(currDir, sourceMapLine);
+      fs.readFile(mapUrl, 'utf8', function (err, mapContents) {
+        if (err) {
+          logger.warn('Error loading source map at ' + mapUrl);
+          return cb(null, { hash: hash, lines: lines });
+        }
+        var compilationInfo = {};
+        var sourceString = MAP_SOURCES_REGEX.exec(mapContents);
+        if (sourceString && sourceString.length > 1) {
+          var sources = JSON.parse(sourceString[1]);
```
Comment (ofrobots) [inline lib/scanner.js]: Need a try catch. `sourceString[1]` could be malformed.

File: lib/scanner.js [hunk 3]
```diff
@@ -134,31 +144,56 @@ function findJSFiles(baseDir, callback) {
...
+          });
+        }
+        return cb(null, {
+          hash: hash,
+          lines: lines,
+          compilationInfo: compilationInfo
+        });
+      });
+    } else {
+      return cb(null, { hash: hash, lines: lines });
```
Comment (ofrobots) [inline lib/scanner.js]: Reverse the `if` condition to get this simpler case out of the way first.

File: lib/scanner.js (hunk 1 above)
Comment (matthewloring) [inline lib/scanner.js]: Done.

File: lib/scanner.js (hunk 2 above)
Comment (matthewloring) [inline lib/scanner.js]: Done.

File: lib/scanner.js (hunk 3 above)
Comment (matthewloring) [inline lib/scanner.js]: Done.

**Comment by matthewloring** [comment]
@ofrobots Ready for another look. There is now an error returned when the single source file in multiple compilation targets case is detected. This case will be addressed in the next PR.

**Comment by matthewloring** [comment]
Closing for now.

**Pull request closed without merging**